### PR TITLE
Add support for DISTINCT statement

### PR DESCRIPTION
**Pull request by bingo-soft**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | 

#### Summary

The same functionality as in Doctrine/ORM. New method `distinct($flag = true)` in `QueryBuilder` class and new test `testSimpleSelectWithDistinct` in `QueryBuilderTest` are provided.

### CODE REVIEW

**Comment by greg0ire** [comment]
That functionality is already available by using `select('DISTINCT u.id')` correct? Is the point to be able to make it conditional easily? Can you provide a code sample that shows in what situation this will make your life easier?

**Comment by bingo-soft** [comment]
@greg0ire. There is a number of arguments for providing this functionality.

First of all, in SQL `DISTINCT` is a separate operation, similar to `GROUP BY`, but with a specific focus on eliminating duplicates. So, if we have `groupBy` in API, why not to provide a more granular `distinct` method?

Secondly, `DISTINCT` is a row operation, not a column function, so if you write higher level query language on top of Doctrine DBAL, which allows users to specify returned fields on the fly, it would be easier to apply `distinct=true` once, then to inject this logic into the part of the algorithm responsible for handling fields. So, two lines of code in Doctrine DBAL would save numbers of lines of client code (looping through all fields, injecting DISTINCT to the first fields; if client does not provide the list of fields, than adding DISTINCT to `*`). So, what is expected to be rather a row operation, becomes a column operation in DBAL client code.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: I don't think this cast is needed.
```suggestion
        $this->sqlParts['distinct'] = $flag;
```

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]:
```suggestion
    final public function distinct(bool $flag = true) : self
```

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: This is already supposed to be a boolean, and if happens not to be, the bug might still go undetected. If you feel unsafe about it being a boolean, then check that and throw if it isn't, or introduce a private accessor to do the type check. If you feel safe, then simply do this:

```suggestion
        $query = 'SELECT ' . ($this->sqlParts['distinct'] ? 'DISTINCT ' : '') .
```

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: I'd move it one line below, it looks like there this array is ordered with the same order in which you would find these keywords in an SQL query.

**Review by greg0ire — CHANGES_REQUESTED** [review]
> Also, is develop the correct target branch when there are no BC breaks?

Please address this

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: In what scenario is that parameter going to be useful?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (bingo-soft) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Looks like that. You are right

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (bingo-soft) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Agree with `self` part, but not sure of `final` specificator. It would be the only one `final` method in QueryBuilder.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (bingo-soft) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Do you mean, that there is no sense to mention this parameter in the docstring, since it is self evident? And by the way, as for `develop` branch, should I make a new merge request to the master branch to fix this?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Yeah but it is one of the most recent ones, right? We can wait for the advice of another maintainer if you want, I will not merge this without the approval of some other maintainers anyway.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (greg0ire) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: No I mean, maybe we should remove the parameter entirely?

> And by the way, as for develop branch, should I make a new merge request to the master branch to fix this?

You can edit your PR to target master, and rebase your branch on it: `git rebase --onto master origin/develop develop` (you should probably not have named your local branch `develop` BTW, it makes this a bit confusing :P)

**Comment by greg0ire** [comment]
You should probably add something to https://github.com/doctrine/dbal/blob/master/docs/en/reference/query-builder.rst

File: lib/Doctrine/DBAL/Query/QueryBuilder.php
Comment (bingo-soft) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Ups, I've never done that magic yet and made another fork for another PR. So, I think I will close this one and start another PR, if you don't mind. And I will take into account all you suggestions, probably except `final` key word

**Pull request closed without merging**
